src/thor/centroid.cc

```diff
@@ -0,0 +1,248 @@
+      PathIntersection{baldr::kInvalidGraphId, baldr::kInvalidGraphId, location_count_};
+
+  // tell dijkstras we want to track the origin paths separately/concurrently
+  multipath_ = true;
+
+  // compute the expansion
+  Dijkstras::Compute(locations, reader, costings, mode);
+
+  // create the paths from the labelset
+  return FormPaths(locations, bdedgelabels_, reader, centroid);
+}
+
+// this is fired when the edge in the label has been settled (shortest path found) so we need to check
+// our intersections and add or update them
+thor::ExpansionRecommendation Centroid::ShouldExpand(baldr::GraphReader& reader,
+                                                     const sif::EdgeLabel& label,
+                                                     const thor::InfoRoutingType) {
+  // TODO: we should quit earlier if finding a centroid isnt working out
+
+  // TODO: refactor dijkstras a bit to get the tile and send it to us so we dont have to
+
+  // grab the opposing edge if you can
+  const baldr::GraphTile* tile = nullptr;
+  baldr::GraphId opp_id;
+  if (const auto* node = reader.nodeinfo(label.endnode(), tile)) {
+    opp_id = tile->header()->graphid();
+    opp_id.set_id(node->edge_index() + label.opp_index());
+  }
+
+  // see if we have seen this edge before
+  PathIntersection intersection(label.edgeid(), opp_id, location_count_);
+  auto found = intersections_.find(intersection);
+
+  // if not we create the record
+  if (found == intersections_.end()) {
+    found = intersections_.insert(std::move(intersection)).first;
+  }
+
+  // update the record to include this path
+  bool is_centroid = found->AddPath(label.path_id());
+
+  // TODO: we should probably reject certain road classes as a centroid if desired, if you wanted to
+  // actually drive these paths it doesnt make sense to meet other drivers on a limited access road
+
+  // TODO: prune these when they are outside of a reasonable bounding box, if that leads to failure
+  // drop the bounding box and dont prune
+
+  // quit as soon as we find a centroid, otherwise always continue looking
+  if (is_centroid) {
+    best_intersection_ = *found;
+    return thor::ExpansionRecommendation::stop_expansion;
+  }
+  return thor::ExpansionRecommendation::continue_expansion;
+}
+
+// tell the expansion how many labels to expect and how many buckets to use
+void Centroid::GetExpansionHints(uint32_t& bucket_count, uint32_t& edge_label_reservation) const {
+  // TODO: come up with a heuristic based on the expansion we expect to have to do (input locations)
+  bucket_count = 20000;
+  edge_label_reservation = 500000;
+}
+
+// deallocate and prepare for next request
+void Centroid::Clear() {
+  intersections_.clear();
+  Dijkstras::Clear();
+}
+
+// walk edge labels to form paths for each location to the centroid
+template <typename label_container_t>
+std::vector<std::vector<PathInfo>>
+Centroid::FormPaths(const google::protobuf::RepeatedPtrField<valhalla::Location>& locations,
```

Comment:
nothing interesting here, just loop over the locations and recover their paths individually